socket.gaierror: [Errno 8] nodename nor servname provided, or not known

Dear James,

first things first: Thanks a stack for conceiving and maintaining this excellent package. It is tremendously helpful for conducting full system tests.

We are currently in the process of giving some love to an abandoned Python project over at [1,2], where we would like to use `smtpdfix` to verify email delivery within the test harness.

When starting with `smtpdfix`, we quickly ran into #89. However, we were happy to find #95, which already removed that feature. Other than the potential hiccup for some users where `socket.gethostbyname()` croaks, we believe it will be a great enhancement because the service will be available for consumption instantly, without going through key material and certificate generation, right?

So, we already recognized you were preparing version 0.4.0, at #144. May we humbly ask if you could publish version 0.4.0 to PyPI to make it available for pulling it into downstream projects? We think it would be a great benefit for the community. Please let us know if there are some blockers where we may be able to help out.

With kind regards,
Andreas.

[1] https://github.com/isarengineering/SecPi
[2] https://github.com/SecPi/SecPi/issues/120

---

**Describe the bug**
When the library is initializing, it croaks like:
```python
        # Generate public certificate
        hostname = socket.gethostname()
>       ip = socket.gethostbyname(hostname)
E       socket.gaierror: [Errno 8] nodename nor servname provided, or not known

.venv/lib/python3.10/site-packages/smtpdfix/certs.py:35: gaierror
```

**To Reproduce**
It happens on slightly misconfigured systems where the designated hostname does not resolve to an IP address.
```python
>>> import socket
>>> socket.gethostbyname(socket.gethostname())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```

**Expected behavior**
The library should work instantly by starting an SMTP server on port 8025.

**Environment (please complete the following information):**
 - macOS 10.15.7
 - Python 3.10
 - smtpdfix 0.3.3

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -13,6 +13,7 @@ Release date: TBD
 - Previously the `AuthController._get_ssl_context.resolve_context` method would return `None` if the `file_` argument was `None`. This will now raise an error instead. The method has been renamed `_resolve_context` to clarify that it should not be considered part of the public API.
 - Adds PyPy 3.9 to CI and `tox.ini` along with appropriate entries in the `minimum\requirements.txt` to reflect that PyPy 3.9 requires cryptography >= 37.0.0. [Issue #166](https://github.com/bebleo/smtpdfix/issues/166)
 - Mark the Authenticator class as being abstract by setting the metaclass to be `abc.ABCMeta` and decorating the methods with `@abc.abstractmethod`
+- On systems where the hostname does not resolve to an IP address the key and certificate will now generate. [Issue #195](https://github.com/bebleo/smtpdfix/195) reported by [Andreas Motl (@amotl)](https://github.com/amotl).
 
 ## Version 0.3.3
 

--- a/smtpdfix/certs.py
+++ b/smtpdfix/certs.py
@@ -35,7 +35,6 @@ def _generate_certs(path: Union[Path, str],
 
     # Generate public certificate
     hostname = socket.gethostname()
-    ip = socket.gethostbyname(hostname)
     subject = Name([NameAttribute(NameOID.COMMON_NAME, "smtpdfix_cert")])
     alt_names = [
         DNSName("localhost"),
@@ -44,8 +43,18 @@ def _generate_certs(path: Union[Path, str],
         IPAddress(ip_address("127.0.0.1")),
         IPAddress(ip_address("0.0.0.1")),
         IPAddress(ip_address("::1")),
-        IPAddress(ip_address(ip)),
     ]
+
+    # Because on misconfigured systems it's possible to have a hostname that
+    # doesn't resolve to an IP we catch the error and skip adding it to the
+    # list of altnames. (issue #195)
+    try:
+        ip = socket.gethostbyname(hostname)
+        alt_names.append(IPAddress(ip_address(ip)))
+    except socket.gaierror as err:
+        log.info(f"{hostname} failed to resolve to ip")
+        log.error(err.strerror)
+
     # Set it so the certificate can be a root certificate with
     # ca=true, path_length=0 means it can only sign itself.
     constraints = BasicConstraints(ca=True, path_length=0)

--- a/tests/test_smtpdfix.py
+++ b/tests/test_smtpdfix.py
@@ -1,4 +1,6 @@
+from pathlib import Path
 from smtplib import SMTP
+from typing import Any
 
 from smtpdfix import SMTPDFix
 
@@ -7,3 +9,19 @@ def test_smtpdfix(msg):
     with SMTPDFix() as server, SMTP(server.hostname, server.port) as client:
         client.send_message(msg)
         assert len(server.messages) == 1
+
+
+def test_misconfigured_socket(monkeypatch, tmp_path_factory):
+    # As reported in #195 a misconfigured system will raise an error if
+    # the hostname won't resolve to an IP address
+    def raise_GAIError(*args: Any) -> None:
+        from socket import gaierror
+        raise gaierror("[Errno 8] nodename nor servname "
+                       "provided, or not known")
+    monkeypatch.setattr("socket.gethostbyname", raise_GAIError)
+
+    from smtpdfix.certs import _generate_certs
+    path = tmp_path_factory.mktemp("certs")
+    _generate_certs(path)
+
+    assert Path.joinpath(path, "cert.pem").is_file()

